### ISSUE TITLE
POST request and header creation questions

### ISSUE THREAD

**Issue by nidoro** [warmcat/libwebsockets#1358]
Hello,
I'm trying to use libwebsockets to communicate with a xml-rpc server. The message format is pretty straight forward, I basically need to send a post request with a few required headers and an xml content (example in the specs [xmlrpc.scripting.com/spec.html](url)).

So, I'm the client writing a POST request to a working server. Right now the server returns a Bad Request error.

1. Is there a way to print the request just before it is sent, with headers and content, so I can inspect it and see it I screwed it up somewhere?
2. At the time that the `LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER` callback is called, did libwebsockets already write some common headers? Or do I have to do everything myself?
3. I read something in the docs about having to handle line breaks in the headers myself. Do I have to do it even when using `lws_add_http_header_by_token(...)`?
4. When writing headers, do I have to use a buffer with `LWS_PRE` padding, or is this just for `lws_write(...)`?

**Comment by lws-team** [comment]
To get a general idea look here

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-client/minimal-http-client-post

For server you can see everything in endless detail by upping the logging.  For client it just adds your headers to whatever GET or POST you requested, plus mandatory headers produced from your request depending on http version... to confirm it you can use eg `lwsl_hexdump_notice()` here:

https://libwebsockets.org/git/libwebsockets/tree/lib/core/output.c#n27

But in the end it's up to the server and intermediaries (who may munge the headers) what it sees and how it feels about that.  Debugging from server side is best if the server has an objection.  Because its reason for objecting may be, eg, a policy that a specific special auth header must be present.  You can't figure out its arbitrary policy from the client side.

Lws supports http/2 for a while, headers go out in their own frame using hpack, the days you can shove your own CRLF are gone.  Use the header helpers as shown in the examples so it can work on both h1 and h2 the same.  You must use LWS_PRE padding because the headers may go out on h2, with its own frame header prepended...

**Comment by nidoro** [comment]
It seems like my problem is that I need to send both headers and content in one go, but the example given sends the content after the headers (at least this is what is implied [in the lws_client_http_body_pending doc](https://libwebsockets.org/lws-api-doc-master/html/group__service.html)). How can I write the body in the handshake message? I tried calling lws_write just after adding my headers but it seems like it overwrites the headers or something. Maybe it is stupid what I'm trying to do but this is what my code looks like:

```
if (LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER == reason) {
        println("LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER");
        
        uint8_t** up = (uint8_t**) in;
        uint8_t* uend = *up + len - 1;
        
        const char contentType[] = "text/xml";
        const char userAgent[] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36";
        
        if (lws_add_http_header_by_token(
                socket, 
                WSI_TOKEN_HTTP_USER_AGENT, 
                (uint8_t*) userAgent, strlen(userAgent), up, uend
           )) {
            return 1;
        }
        
        if (lws_add_http_header_by_token(
                socket,
                WSI_TOKEN_HTTP_CONTENT_TYPE,
                (uint8_t*) contentType, strlen(contentType), up, uend
           )) {
            return 1;
        }
        
        lws_add_http_header_content_length(socket, 84, up, uend);
        
        //lws_client_http_body_pending(socket, 1);
        //lws_callback_on_writable(socket);
        
        char buffer[LWS_PRE + 2048] = {};
        const char xml[] = 
        R"(<?xml version="1.0"?>
                <methodCall>
                <methodName>help</methodName>
                </methodCall>
         )";
        
        strcpy((char*)*up, xml);
        println((char*)*up);
        lws_client_http_body_pending(socket, 0);
        lws_write(socket, (unsigned char*) *up, sizeof(xml), LWS_WRITE_TEXT);
}
```

**Comment by lws-team** [comment]
>  I need to send both headers and content in one go

That's really unlikely to be the problem, since TCP/IP is able to be fragmented by any intermediary, if the server can't handle that it would be fragile to the point of uselessness.

Lws supports h2, in that headers go out in a completely different frame than the payload, which has no relationship to the tcp packets carrying the frame.  So they must be separated and that's not going to change.

What led you to think that this might be "the problem"?

**Comment by nidoro** [comment]
Well, nevermind.

> What led you to think that this might be "the problem"?

I was sending my request to a little server I wrote in python just to inspect what was coming out of my lws client request and I was only getting the headers, with no content, and after that the connection was closed, so I though 'well, maybe it's closing because the message has no body', but that was just me talking about something I don't know.

The root of all my problems was very simple: the URI was empty, without '/', because I thought that was the way it should be. I guess that's what I get for not reading specs carefully.

Thanks for the help! And for the library!
